### ISSUE TITLE
Throughput scaling - Jupiter - Qwen3-32B Dense model

### ISSUE THREAD

**Issue by Vmjkom** [OpenEuroLLM/Taskboard#217]
In preparation for the flagship model training, we need to ensure satisfactory throughput with a dense model. This issue shall get updated as the details get figured out.

## Goal
Using the [https://huggingface.co/Qwen/Qwen3-32B](https://huggingface.co/Qwen/Qwen3-32B) configuration get preliminary results for throughput scaling up to X amount of nodes.

## Target
What can be considered "satisfactory" throughput? Look for previous scaling results done by OpenSci folks and others.

### Considerations

- What batch size to use?
- Best model parallelism configuration

**Comment by JeniaJitsev** [comment]
Satisfactory is a result close to 400 TFLOPS/s/GPU. For 4k context length, we should see ca. 1500-2000 Tokens/s/GPU for 32B model scale. Those are acceptable numbers. Above that of course even better (eg 450-500 TFLOPS, 2000-2500 Tokens/s/GPU), probably hard to get given current state of the machine.

**Comment by Vmjkom** [comment]
Current target should be to first match Marianna's 400 TFLOPS/GPU which should be achievable from 4 nodes and up. 
Model parallelism config, GBS and MBS for 4 nodes for that run:
| TP | PP | GBS | MBS |
|--------|--------|--------|--------|
| 4 | 2 | 25 | 2 |


## Questions
* Should we use FP8?
* How to best make use of the NVLink-C2C (chip-to-chip) connection? Should we look closer at CPU offloading techniques to get bigger MBS or to relieve us from using PP=2 and/or TP=4?

**Comment by Vmjkom** [comment]
## Update 16.6.2026

Current best performing config is

| TP | PP | VPP | GBS | MBS |
|--------|--------|--------|--------|--------|
| 4 | 2 | 2 | 256 | 2 | 

Where the average throughput is

| TFLOPS | TKS/GPU |
|--------|--------|
| 385 | 1850 | 


### Next steps

Look at increasing MBS to 4 through heavier use of recompute activations and other memory saving techniques. The hope here is better gemm efficiency and less pipeline bubble. 
Other than increasing the MBS i will look at FSDP vs the model parallelism config we have now as well as FP8 performance gains.

- [ ] FSDP
- [ ] FP8
- [x] Sweeps for recompute and mbs4

## Notes

Turning off GPU direct RDMA with `export NCCL_NET_GDR_LEVEL=0` and disabling slurm gpu bind with `--gpu-bind=none` results in performance increase of about 20 TFLOPS from their respective default values. **TO BE CONFIRMED**

**Comment by Vmjkom** [comment]
# Update 23.6.2026

JUPITER is offline and the weak-scaling results up to 1024 nodes can be seen here in this [WandDB report](https://api.wandb.ai/links/openeurollm-project/e1tit83m).

## Best configuration

| TP | PP | VPP | MBS |
|--------|--------|--------|--------|
| 4 | 2 | 2 | 2 |

extra parameters:
```
tp_comm_overlap
cross_entropy_loss_fusion
```

Adding TP comms overlap resulted in performance increase of about 4% respectively to the "baseline" of 385 TFLOPS mentioned in the last update. With these enabled TFLOPS at 4 nodes is **~403.**

<img width="5561" height="2921" alt="Image" src="https://github.com/user-attachments/assets/2dd3e817-cfe4-4e31-9e8a-697c24e2d0ae" />

## Weak scaling

<img width="5561" height="2921" alt="Image" src="https://github.com/user-attachments/assets/1b516cf3-e615-4f18-9e98-34c1a1e05c13" />

Performance drops to around 370 TFLOPS on the final node count of 1024. 

## Notes

### Environment variables and params

I found there be some redundant environment variables in the [autoexp slurm config for JUPITER](https://github.com/OpenEuroLLM/oellm-autoexp/blob/main/config/slurm/jupiter.yaml#L21-L49) and some which might be bad for performance in large scale runs. I'll be PR'ing my own config with validated environment variables once JUPITER comes back up. I'll also then give my take which ones should be set and why.

OMP_NUM_THREADS=1 caused GPU's to run idle, so setting it to a higher value helps performance, what that value is still a question, but i've set it to the full CPU core count of 288. Afaik, this creates 64 virtual threads for each task, which is probably overkill, so tweaking this needs to be looked at.

Another thing i found to be needed for optimal performance is setting slurm's gpu binding off with `--gpu-bind=none`. I don't know why this is the case, it hasn't been necessary in other clusters. Pytorch binds GPU's with processes via local rank, but it might be the case that NCCL gets suboptimal affinities from SLURM.

#### UCX

[Some JSC recommended UCX variables, that haven't been tried ](https://apps.fz-juelich.de/jsc/hps/jupiter/buildup.html#tuning-for-large-scale-execution).

### Recomputation

In order to increase MBS for GEMM efficiency among other things, experiments with activation recomputing we're conducted, but these didn't result in performance increase.

### Power cap

The compute node caps the power wattage of the GH200 to 680W. By default, 100W are reserved to CPU, but this can be lowered to give more power to the GPU.
https://apps.fz-juelich.de/jsc/hps/jupiter/buildup.html#currently-supported-options

### Node topology
I saw some discrepancies between identical (same script/config) and my hypothesis for this is suboptimal node allocations, where node's might be in different dragonfly groups (this phenomena was noticed on 16 nodes), or otherwise far away from each other in the topology. We could ask JSC to provide us with a good topology for the training run if we can confirm huge performance loss due to this. [Relevant jsc docs](https://apps.fz-juelich.de/jsc/hps/jupiter/batchsystem.html#generic-resources-features-and-topology-aware-allocations).

### I/O
These runs we're only ran with a single dataset and .bin and .idx files. The files we're located in $PROJECT, but $SCRATCH is meant for high I/O work, so we need to benchmark against this. Running with the full dataset means multiple filepaths and reads and writes.

#### Dataloader
Interestingly, i had slightly worse performance with --mock-data, than with a real dataset. However, perf with --mock-data should be used as a baseline for benchmarking dataloader for the full-dataset run.
So far, the number of dataloader workers has been 2, but this might need to be increased as the dataset grows in size and number of files.
## Next steps

#### Megatron core version

I found there to be no difference in performance between 0.15 and 0.16 with the parameters set in weak-scaling runs, but 0.17 was worse off. There are differences in what parameters are available in each of the versions, but only 0.17 seemed to have different values for defaults in the parameters seen in all versions.

## Next steps 

I ran out of time before i could profile and optimize throughput for the large node counts where performance dips, but there's a few things we might do for that still.

- FSDP is a worthwhile experiment because of GH200's C2C highspeed nvlink (link between CPU and GPU). It's heavier on the comms side and hasn't been widely used in very large scale training afaik.
- In one of the experiments FP8 gave a performance and the MBS could have been increased from 2 to 4 probably, but this needs to be confirmed. Also a general decision on whether FP8 should be used needs to be made.

I would need to collect more comms timings and profile runs to get a better understanding what should be done to have better scaling, or even better general throughput. Some parameters suggested by an LLM:

- num_distributed_optimizer_instances 
- `--account-for-embedding-in-pipeline-split --account-for-loss-in-pipeline-split`
- DDB bucket size tuning for better bandwidth

**Comment by zeppelinche-cmd** [comment]
Continuing work in https://github.com/OpenEuroLLM/Taskboard/issues/334, we consider the scope of this one closed.

**Comment by otaub** [comment]
I'm going to do
strong scaling runs for
GBS: 1k, 2k, (3k), 4k
Nodes: (4 16 32) 128 256 ,512, 1024
in terms of TFLOPS/GPU and tok/s
to obtain the same graphs as @e-cozzolino for the 32B model on Jupiter, see message from May 14th, 10:12 in T4.1. (lower priority in brackets).

**Comment by Vmjkom** [comment]
# Final optimization


## 8, 16 and 64 nodes profiling 

We see the biggest drop in performance going from 8 to 16 nodes. The next big drop is from 32 to 64 nodes. The suspected bottleneck is the comms. To investigate further i'll do a few runs on the aforementioned node counts in which i'll enable timer logging on the megatron side, profile the runs and look at NCCL logs for message sizes.

- [x] profile 8, 16 and 64 node count runs and see what's up

## 64 nodes

### FP8 vs BF16
In smaller a smaller scale experiment FP8 gave a very slight performance boost. I'll test this on 64 nodes to get a better idea. 

- [x] FP8 vs BF16

## 1024 nodes

In order to get a proper signal, we do the following DDP 
**Do the following in one allocation!**

The prime suspect for throughput dropping in larger scale is the data parallel communication. To reduce latency one can tune the amount/size of gradients in DDP reduce calls with `--ddp-bucket-size`. 

- [x] Find out the default value of --ddp-bucket-size and make a grid of values to profile.

Another thing we can do for the DDP domain is increase `--num-distributed-optimizer-instances`, which splits the optimizer comms to smaller groups. This reduces their size, but might be worthwhile for overlap and less congestion.

- [x] Try --num-distributed-optimizer-instances values 2,4,8

### SHARP

Previous experiments show slight performance increase. First we need to make sure it works as intended on JUPITER - then measure performance.

- [ ] Validate SHARP performance

## Power cap, THP and vboost

Jupiter offers several ways to manage resources for the hardware through a slurm plugin. [The user can enable, disable and otherwise tune parameters with --comment sbatch parameter](https://apps.fz-juelich.de/jsc/hps/jupiter/buildup.html#currently-supported-options).

### GPU Power cap and clock frequency

As we know, the GH200 chip on Booster is power capped to 680W where 100W goes to the CPU by default and the rest to the GPU. There are options to adjust the power cap for both GPU and CPU; [it seems to be possible to lower CPU's power from 100W](https://apps.fz-juelich.de/jsc/hps/jupiter/buildup.html#gh200-superchip-power-cap) in which case the excess would go to the GPU. 

- [x] Tune transfer of power from CPU to GPU through the --comment sbatch parameter.

### GPU clocks
The same document entails that the default GPU clocks might not be optimized for compute intensive workloads like our LLM pretraining. 

- [x] Profile performance with higher GPU clock frequencies

### THP - Transparent huge pages

[Apertus reported THP not working optimally for their 1024 node LLM pretraining](https://arxiv.org/html/2604.12973v2). In JUPITER THP is enabled by default.

- [ ] investigate whether disabling it with `--comment=THP=never` offers differences in throughput.

### Nvidia vboost

Apertus also reported performance increase when allowing more compute power directed to the GPU at CPU's expensive with [vboost](https://man.archlinux.org/man/nvidia-smi.1.en#nvidia_smi~169). Setting the value is permission restricted for the user on login nodes.
```
[komulainen1@jpbl-s02-02 jupiter]$ nvidia-smi boost-slider --vboost=1
Failed to set vboost slider with value 1 for GPU 0 : Insufficient Permissions
Failed to set control for vboost slider for requested devices : Insufficient Permissions
```

- [x] Investigate whether vboost is enabled by the admins and tune to higher values.

**Comment by Vmjkom** [comment]
## Nvidia vboost

(.venv) bash-5.1$ nvidia-smi boost-slider --vboost=1
Failed to set vboost slider with value 1 for GPU 0 : Insufficient Permissions
Failed to set control for vboost slider for requested devices : Insufficient Permissions

Same problem with compute node. We should reach out to JSC whether we could access this via slurm option, like Apertus did.

**Comment by Vmjkom** [comment]
# Update 24.7.2026

## SHARP

Currently Nvidia sharp isn't working. I think it's due to something changing during the downtime, as i had it working before. I have contacted JSC support and we'll hopefully get it sorted, as i think the problem is only resolvable by a slurm admin who needs to enable the Aggregation Manager (sharp_am seen in the code). [SHARP should be pursued for performance gains with DP sizes](https://docs.nvidia.com/nemo/megatron-bridge/0.5.0/performance-guide.html#performance-at-scale).
```
[default2]:[jpbo-039-13][Jul 24 06:20:53 418410][RDMA_SR][1286392][error] - poll failed due to poll_timeout=3000.000000, stop
[default2]:[jpbo-039-13][Jul 24 06:20:53 418488][RDMA_SR][1286392][error] - Poll failed
[default2]:[jpbo-039-13][Jul 24 06:20:53 418500][RDMA_SR][1286392][error] - Failed to connect
[default2]:[jpbo-039-13][Jul 24 06:20:53 420389][RDMA_SR][1286392][error] - rdma_resolve_addr failed with error: -1
[default2]:[jpbo-039-13][Jul 24 06:20:53 420426][RDMA_SR][1286392][error] - rdma_resolve_addr failed with error: -1
[default2]:[jpbo-039-13][Jul 24 06:20:53 420443][SR     ][1286392][error] - unable to query AM service record(AM query)
[default2]:[jpbo-039-13][Jul 24 06:20:53 420461][GENERAL][1286392][error] - Could not query AM address, error: -52
[default2]:[jpbo-039-13][Jul 24 06:20:53 420475][GENERAL][1286392][error] - unable to connect to AM
[default2]:[jpbo-039-13][Jul 24 06:20:53 420489][GENERAL][1286392][warn ] - SHARPD_OP_CREATE_JOB failed with status: 52
[default2]:[jpbo-039-13:0:1285361 unique id 9226306752374689105][2026-07-24 06:20:53] ERROR No Aggregation Manager (sharp_am) detected in sharp_create_job.
[default2]:
[default2]:[jpbo-039-13:0:1285361 - context.c:738][2026-07-24 06:20:53] ERROR sharp_create_job failed: No Aggregation Manager (sharp_am) detected(-52)

``` 

## GPU Power cap and clock frequency

I verified that GPU power cap cannot be extended from the 680W, even if CPU power cap is lowered to 50W. It is limited by the overall module power cap, which is 680W.
GPU clock's can be locked onto a number of values, with the maximum being (it seems) 1980Mhz. This can be achieved with 
```
#SBATCH --comment=NVIDIA_GPU_CLOCKS=1980
``` 
Logs and script for testing these things out are in `/e/scratch/e-sta-openeurollm/komulainen1/power_experiments`

### Experiments on 16 nodes

The following [Wandb table](https://wandb.ai/openeurollm-project/qwen3-32B-dense-scaling-jupiter/table?nw=52mt4hl6gy) shows that locking GPU clocks to the maximum value and increasing DDP bucket size 2x from the default resulted in ~17 and ~12 increase in TFLOPS, respectively. These are singular results and the effects don't compound. Increasing the number of optimizer instances unsurprisingly doesn't have any effect as the DP size is relatively small.

**Comment by Vmjkom** [comment]
# Update 03.08.2026

## Smaller updates to previous points

### Clocks

Locking the GPU clock to 1980Mhz with `#SBATCH --comment=NVIDIA_GPU_CLOCKS=1980` brings around 16 TFLOP increase on 16 nodes.

### DDP bucket size

After further experimentation on larger node counts and varying values for the bucket size, experiments show no throughput benefit in neither case of increasing or decreasing from the default value. DDP bandwidth seems to be OK. [Wandb panel of 512 node runs](https://wandb.ai/openeurollm-project/qwen3-32B-dense-scaling-jupiter/table?nw=90rme9dxwbm).

### SHARP

JSC support got back to me on this issue and said that SHARP is indeed not working due to changes that happened on the system after maintenance and further noted that they are looking at fixing this, but not in the immediate future.

## Lower GBS

Weak scaling GBS was unrealistic on node counts we're targeting for pretraining, so i started experimenting with lower GBS. Most of current optimization has been done on a GBS of 8192. 

### Pipeline imbalance

Profiles show that with our PP=2, the later stage of the physical PP stages was always slower than the first, which meant that PP comms scaled badly starting from 128 nodes. 

#### Decrease virtual pipeline parallelism

Baseline is 2 layers per virtual pipeline stage, which resulted in VPP of 16. Decreasing this by setting `--num-layers-per-virtual-pipeline-stage=8 `helped comms scaling and improved throughput, but not completely. With the aid if ChatGPT i then implemented a custom pipeline layout for the devices with `--pipeline-model-parallel-layout "'Et*9|t*8|t*8|t*8|t*8|t*8|t*8|t*7L'"`. This made PP comms scale almost perfectly from 4 to 512 nodes.

#### DP comms

The remaining visible "bottleneck" is the grad sync at the end of a training step. The DP comm here does not overlap with computation very well at the moment, but the bandwidth is fine as seen with DDP bucket size experiments.

## Results

I manage to match previous weak scaling results with 8192 GBS strong scaling. This was with `--num-layers-per-virtual-pipeline-stage=8`. 

# Next steps

Strong scaling with the following GBS
- [ ] 1024
- [ ] 2048
- [ ] 4096
- [ ] 8192

**Comment by Vmjkom** [comment]
# Update 20.08.2026

A report on speed tests done by @kpoeppel can be seen [here](https://github.com/OpenEuroLLM/oellm-autoexp/blob/prod/oellm_32b_dense/config/experiments/oellm_32b_dense/SPEED_RESULTS.md#1-pp4-is-the-optimum--more-pp-is-not-better).

## 1024 node fix

Previous crashes and timeouts on 1024 nodes we're due to to torchrun's dynamic rendezvous backend. Falling back to static fixes the issue. This is achieved by setting `--rdzv_backend static`, instead of c10d.

## FP8

### Stability
@kpoeppel has ongoing runs regarding this [here](https://wandb.ai/openeurollm-project/oellm_32b_dense_stability_speed_checks).

### Performance vs BF16

Currently we can assume that FP8 Performance gains to be substantial across a wide range of node counts. Korbinian reports performance of around [410 TFLOPS](https://github.com/OpenEuroLLM/oellm-autoexp/blob/prod/oellm_32b_dense/config/experiments/oellm_32b_dense/SPEED_RESULTS.md#3-fp32-gradient-accumulation-is-free-at-512-nodes-costs-79-at-1024) on 512 nodes. This was achieved with `--fp8-recipe 'tensorwise'`. [Precision config](https://github.com/OpenEuroLLM/oellm-autoexp/blob/prod/oellm_32b_dense/config/backend/megatron/oellm_32b_dense/precision/fp8.yaml).
I myself got around 50 TFLOPS improvement over my BF16 baseline.

<img width="1920" height="1017" alt="Image" src="https://github.com/user-attachments/assets/87cbdd06-bb21-432c-bb41-bfcc01adaf83" />

## PP 4

@kpoeppel [reports PP=4 being more performant over PP=2 in 512-1024 node range](https://github.com/OpenEuroLLM/oellm-autoexp/blob/prod/oellm_32b_dense/config/experiments/oellm_32b_dense/SPEED_RESULTS.md#1-pp4-is-the-optimum--more-pp-is-not-better).
## SHARP

Even with the most recent Mellanox update on the 12th, Sharp remains disabled by the admins due to recent changes to SHARP by Nvidia.

## Full dataset I/O

From a quick test looking at purely TFLOPS, reading and writing from the whole dataset brings about no issues in throughput that we've seen on other systems. The run below is a comparison between `--mock-data`, 

<img width="3264" height="636" alt="Image" src="https://github.com/user-attachments/assets/4ecd3cb7-e761-41de-acb4-59fdc96f8109" />
